api/app/utils/enphase_utils.py

```diff
@@ -0,0 +1,126 @@
+from quartz_solar_forecast.forecast import predict_tryolabs
+from quartz_solar_forecast.data import get_nwp, process_pv_data
+from quartz_solar_forecast.inverters.enphase import process_enphase_data
+
+def get_enphase_access_token(auth_code):
+    client_id = os.getenv("ENPHASE_CLIENT_ID")
+    client_secret = os.getenv("ENPHASE_CLIENT_SECRET")
+    redirect_uri = os.getenv("ENPHASE_REDIRECT_URI", "https://api.enphaseenergy.com/oauth/redirect_uri")
+
+    credentials = f"{client_id}:{client_secret}"
+    encoded_credentials = base64.b64encode(credentials.encode("utf-8")).decode("utf-8")
+
+    conn = http.client.HTTPSConnection("api.enphaseenergy.com")
+    headers = {
+        "Content-Type": "application/x-www-form-urlencoded",
+        "Authorization": f"Basic {encoded_credentials}"
+    }
+    payload = f"grant_type=authorization_code&redirect_uri={redirect_uri}&code={auth_code}"
+
+    conn.request("POST", "/oauth/token", payload, headers)
+    res = conn.getresponse()
+    data = res.read()
+    conn.close()
+
+    data_json = json.loads(data.decode("utf-8"))
+
+    if "error" in data_json:
+        raise ValueError(f"Error in getting access token: {data_json['error_description']}")
+
+    if "access_token" not in data_json:
+        raise KeyError(f"Access token not found in response. Response: {data_json}")
+
+    return data_json["access_token"]
+
+def get_enphase_data(enphase_system_id: str, access_token: str) -> pd.DataFrame:
+    api_key = os.getenv("ENPHASE_API_KEY")
+    start_at = int((datetime.now() - timedelta(weeks=1)).timestamp())
+    granularity = "week"
+
+    conn = http.client.HTTPSConnection("api.enphaseenergy.com")
+    headers = {"Authorization": f"Bearer {access_token}", "key": api_key}
+
+    url = f"/api/v4/systems/{enphase_system_id}/telemetry/production_micro?start_at={start_at}&granularity={granularity}"
+    conn.request("GET", url, headers=headers)
+    res = conn.getresponse()
+    data = res.read()
+    conn.close()
+
+    data_json = json.loads(data.decode("utf-8"))
+    return process_enphase_data(data_json, start_at)
+
+def make_pv_data(
+    site: PVSite,
+    ts: pd.Timestamp,
+    access_token: str = None,
+    enphase_system_id: str = None,
+    solis_data: pd.DataFrame = None,
+    givenergy_data: pd.DataFrame = None,
+    solarman_data: pd.DataFrame = None
+) -> xr.Dataset:
+    live_generation_kw = None
+
+    if site.inverter_type == "enphase" and access_token and enphase_system_id:
+        live_generation_kw = get_enphase_data(enphase_system_id, access_token)
+    elif site.inverter_type == "solis" and solis_data is not None:
+        live_generation_kw = solis_data
+    elif site.inverter_type == "givenergy" and givenergy_data is not None:
+        live_generation_kw = givenergy_data
+    elif site.inverter_type == "solarman" and solarman_data is not None:
+        live_generation_kw = solarman_data
+
+    da = process_pv_data(live_generation_kw, ts, site)
+    return da
+
+def predict_ocf(
```

Comment:
im suprised this function is in there? Is this duplicated?

Reply:
> im suprised this function is in there? Is this duplicated?

Since I'm passing the access token for it to persist in the app, it needed to be modified, it's how it's done in the current main branch too

Reply:
I understand you need to pass through the access token. I think there is a neater way of doing this, rather than duplicating the code.

You could
1. Pass through all relevant access tokens, ids e.t.c in a dict in the `ForecastRequest` when calling the api. This could be an environmental variables dict.
2. Set these env vars as necessary for the inverters
3. Call 'run_forecast` from the main library. 

We might need to think slightly carefully how we set the api and dashbaord up, as I think sending secrets thorugh http is considered unsafe. Perhaps we could encode them somehow if necessary, but this can be later.

Reply:
> I understand you need to pass through the access token. I think there is a neater way of doing this, rather than duplicating the code.
> 
> You could
> 
> 1. Pass through all relevant access tokens, ids e.t.c in a dict in the `ForecastRequest` when calling the api. This could be an environmental variables dict.
> 2. Set these env vars as necessary for the inverters
> 3. Call 'run_forecast` from the main library.
> 
> We might need to think slightly carefully how we set the api and dashbaord up, as I think sending secrets thorugh http is considered unsafe. Perhaps we could encode them somehow if necessary, but this can be later.
> 
> Doing it this way thought, should reduce the duplication of code massively

This sounds promising, we could try hashing the token perhaps while passing, but this can be thought of later after getting it working with env vars, thanks for the direction will try this out